plugins/map_matching.hpp

```diff
@@ -0,0 +1,314 @@
+        {
+            return std::make_pair(ClassifierT::ClassLabel::NEGATIVE, 1.0);
+        }
+
+        return label_with_confidence;
+    }
+
+    bool getCandiates(const std::vector<FixedPointCoordinate>& input_coords, std::vector<double>& sub_trace_lengths, Matching::CandidateLists& candidates_lists)
+    {
+        double last_distance = coordinate_calculation::great_circle_distance(
+            input_coords[0],
+            input_coords[1]);
+        sub_trace_lengths.resize(input_coords.size());
+        sub_trace_lengths[0] = 0;
+        for (const auto current_coordinate : osrm::irange<std::size_t>(0, input_coords.size()))
+        {
+            bool allow_uturn = false;
+            if (0 < current_coordinate)
+            {
+                last_distance = coordinate_calculation::great_circle_distance(
+                    input_coords[current_coordinate - 1],
+                    input_coords[current_coordinate]);
+                sub_trace_lengths[current_coordinate] += sub_trace_lengths[current_coordinate-1] + last_distance;
+            }
+
+            if (input_coords.size()-1 > current_coordinate && 0 < current_coordinate)
+            {
+                double turn_angle = ComputeAngle::OfThreeFixedPointCoordinates(
+                                                    input_coords[current_coordinate-1],
+                                                    input_coords[current_coordinate],
+                                                    input_coords[current_coordinate+1]);
+
+                // sharp turns indicate a possible uturn
+                if (turn_angle < 100.0 || turn_angle > 260.0)
+                {
+                    allow_uturn = true;
+                }
+            }
+
+            std::vector<std::pair<PhantomNode, double>> candidates;
+            if (!facade->IncrementalFindPhantomNodeForCoordinateWithMaxDistance(
+                    input_coords[current_coordinate],
+                    candidates,
+                    last_distance/2.0,
+                    5,
+                    Matching::max_number_of_candidates))
+            {
+                return false;
+            }
+
+            if (allow_uturn)
+            {
+                candidates_lists.push_back(candidates);
+            }
+            else
+            {
+                unsigned compact_size = candidates.size();
+                for (const auto i : osrm::irange(0u, compact_size))
+                {
+                    // Split edge if it is bidirectional and append reverse direction to end of list
+                    if (candidates[i].first.forward_node_id != SPECIAL_NODEID
+                     && candidates[i].first.reverse_node_id != SPECIAL_NODEID)
+                    {
+                        PhantomNode reverse_node(candidates[i].first);
+                        reverse_node.forward_node_id = SPECIAL_NODEID;
+                        candidates.push_back(std::make_pair(reverse_node, candidates[i].second));
```

Comment:
using `emplace_back()` here should allow to get rid of the surrounding/inner `make_pair()` call.